From e54e445a15b3519bf45b7b9de1ad19fe42b8608d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fredrik=20M=C3=B6ller?= <fredrik.moller@rt-labs.com>
Date: Tue, 15 Sep 2020 18:22:53 +0200
Subject: [PATCH] rtkernel for Profinet

Changes made to rt-kernel tree from rt-collab
Toolbox 2020.1 as to support the p-net Profinet
stack:
- lwip: Build SNMP server. Discard invalid UDP
  packets. VLAN tag support.
- bsp/xmc48relax: Static IP addressing. Increase
  stack sizes for main() and Ethernet driver
  tasks. Put .bss in different memory section.
---
 bsp/xmc48relax/include/config.h  | 10 +++++-----
 bsp/xmc48relax/src/lwip.c        |  2 +-
 bsp/xmc48relax/xmc48relax.ld     |  2 +-
 lwip/src/apps/Makefile           | 12 ++++++++++++
 lwip/src/apps/snmp/Makefile      | 16 ++++++++++++++++
 lwip/src/core/udp.c              | 12 ++++++++++++
 lwip/src/include/lwip/lwipopts.h | 23 ++++++++++++++++++++++-
 7 files changed, 69 insertions(+), 8 deletions(-)
 create mode 100644 lwip/src/apps/Makefile
 create mode 100644 lwip/src/apps/snmp/Makefile

diff --git a/bsp/xmc48relax/include/config.h b/bsp/xmc48relax/include/config.h
index 850ce674..de1e7de7 100644
--- a/bsp/xmc48relax/include/config.h
+++ b/bsp/xmc48relax/include/config.h
@@ -53,14 +53,14 @@
  *   CFG_LWIP_NAMESERVER        IP address of DNS server
  */
 #define CFG_LWIP_MAC_ADDRESS       {0x1E, 0x30, 0x6C, 0xA2, 0x45, 0x5E}
-#define CFG_LWIP_ADDRESS_DYNAMIC
+#undef CFG_LWIP_ADDRESS_DYNAMIC
 #if defined(CFG_LWIP_ADDRESS_DYNAMIC)
    #define CFG_LWIP_HOSTNAME       "xmc48relax"
 #else
-   #define CFG_LWIP_IPADDR()       IP4_ADDR (&ipaddr, 192, 168, 10, 171)
+   #define CFG_LWIP_IPADDR()       IP4_ADDR (&ipaddr, 192, 168, 137, 4)
    #define CFG_LWIP_NETMASK()      IP4_ADDR (&netmask, 255, 255, 255, 0)
-   #define CFG_LWIP_GATEWAY()      IP4_ADDR (&gw, 192, 168, 10, 1)
-   #define CFG_LWIP_NAMESERVER()   IP4_ADDR (&nameserver, 192, 168, 10, 4)
+   #define CFG_LWIP_GATEWAY()      IP4_ADDR (&gw, 192, 168, 137, 1)
+   #define CFG_LWIP_NAMESERVER()   IP4_ADDR (&nameserver, 0, 0, 0, 0)
 #endif
 
 /* Enable SD Card if defined */
@@ -115,7 +115,7 @@
  * implemented in the user application and is the task that will be
  * started immediately after the kernel has finished initialising.
  */
-#define CFG_MAIN_STACK_SIZE     1280
+#define CFG_MAIN_STACK_SIZE     14374
 
 /* Configure the priority of the main task. */
 #define CFG_MAIN_PRIORITY       10
diff --git a/bsp/xmc48relax/src/lwip.c b/bsp/xmc48relax/src/lwip.c
index f8b7d94a..49780cc4 100644
--- a/bsp/xmc48relax/src/lwip.c
+++ b/bsp/xmc48relax/src/lwip.c
@@ -95,7 +95,7 @@ err_t bsp_eth_init (struct netif * netif)
       .rx_buffers    = 3,
       .tx_buffers    = 5,
       .rx_task_prio  = TCPIP_THREAD_PRIO - 1,
-      .rx_task_stack = 1100,
+      .rx_task_stack = 4000,
       .mac_address   = CFG_LWIP_MAC_ADDRESS,
    };
 #ifdef CFG_RTC_INIT
diff --git a/bsp/xmc48relax/xmc48relax.ld b/bsp/xmc48relax/xmc48relax.ld
index e59a5939..a338c463 100644
--- a/bsp/xmc48relax/xmc48relax.ld
+++ b/bsp/xmc48relax/xmc48relax.ld
@@ -97,7 +97,7 @@ SECTIONS
         event_start = .;
         KEEP (*(.event));
         event_end = .;
-   } > sram
+   } > cdsram
 
    .data : AT (data_start_rom)
    {
diff --git a/lwip/src/apps/Makefile b/lwip/src/apps/Makefile
new file mode 100644
index 00000000..e92bb4ee
--- /dev/null
+++ b/lwip/src/apps/Makefile
@@ -0,0 +1,12 @@
+#********************************************************************
+#        _       _         _
+#  _ __ | |_  _ | |  __ _ | |__   ___
+# | '__|| __|(_)| | / _` || '_ \ / __|
+# | |   | |_  _ | || (_| || |_) |\__ \
+# |_|    \__|(_)|_| \__,_||_.__/ |___/
+#
+# http://www.rt-labs.com
+# Copyright 2013 rt-labs AB, Sweden.
+# See LICENSE file in the project root for full license information.
+#*******************************************************************/
+include $(PRJ_ROOT)/make/subdir.mk
diff --git a/lwip/src/apps/snmp/Makefile b/lwip/src/apps/snmp/Makefile
new file mode 100644
index 00000000..e1383bfe
--- /dev/null
+++ b/lwip/src/apps/snmp/Makefile
@@ -0,0 +1,16 @@
+#********************************************************************
+#        _       _         _
+#  _ __ | |_  _ | |  __ _ | |__   ___
+# | '__|| __|(_)| | / _` || '_ \ / __|
+# | |   | |_  _ | || (_| || |_) |\__ \
+# |_|    \__|(_)|_| \__,_||_.__/ |___/
+#
+# http://www.rt-labs.com
+# Copyright 2013 rt-labs AB, Sweden.
+# See LICENSE file in the project root for full license information.
+#*******************************************************************/
+OBJDIR = ../../../obj/$(ARCH)/$(VARIANT)/$(CPU)
+
+override EXTRA_CFLAGS += -Wno-enum-compare
+
+include $(PRJ_ROOT)/make/subdir.mk
diff --git a/lwip/src/core/udp.c b/lwip/src/core/udp.c
index ce2e3d29..7946cb05 100644
--- a/lwip/src/core/udp.c
+++ b/lwip/src/core/udp.c
@@ -345,6 +345,18 @@ udp_input(struct pbuf *p, struct netif *inp)
       goto end;
     }
 
+    /* rtlaka: discard invalid UDP packets to pass PNIO tester */
+    if ((ntohs(udphdr->len) - UDP_HLEN) != p->len) {
+       /* drop invalid packets */
+      LWIP_DEBUGF(UDP_DEBUG,
+                  ("udp_input: invalid UDP length (%"U16_F" bytes) discarded\n", p->tot_len));
+      UDP_STATS_INC(udp.lenerr);
+      UDP_STATS_INC(udp.drop);
+      MIB2_STATS_INC(mib2.udpinerrors);
+      pbuf_free(p);
+      goto end;
+    }
+
     if (pcb != NULL) {
       MIB2_STATS_INC(mib2.udpindatagrams);
 #if SO_REUSE && SO_REUSE_RXTOALL
diff --git a/lwip/src/include/lwip/lwipopts.h b/lwip/src/include/lwip/lwipopts.h
index c48a69b7..0307f8da 100644
--- a/lwip/src/include/lwip/lwipopts.h
+++ b/lwip/src/include/lwip/lwipopts.h
@@ -49,7 +49,12 @@
 #define LWIP_SOCKET                 1
 #define LWIP_IGMP                   1
 #define LWIP_TCP_KEEPALIVE          1
+#define LWIP_SNMP                   1
+#define SNMP_USE_NETCONN            1
+#define SNMP_USE_RAW                0
+#define MIB2_STATS                  1
 #define SO_REUSE                    1
+#define ETHARP_SUPPORT_VLAN         1
 
 /**
  * LWIP_DHCP_AUTOIP_COOP_TRIES: Set to the number of DHCP DISCOVER probes
@@ -60,20 +65,33 @@
  */
 #define LWIP_DHCP_AUTOIP_COOP_TRIES 2
 
+/* Supported services. Should be 78 for Profinet. See IETF RFC 3418. */
+#define SNMP_SYSSERVICES ((1 << 6) | (1 << 3) | (1 << 2) | (1 << 1))
+
 /* Various memory size options */
 #define MEM_SIZE                    (6 * 1024)
 #define MEMP_NUM_UDP_PCB            6
-#define PBUF_POOL_SIZE              8
+#define PBUF_POOL_SIZE              (8 + 3)
 #define TCP_MSS                     1460 /* Should be 1460 for Ethernet IPv4 */
 #define TCP_SND_BUF                 (2 * TCP_MSS)
+#define SNMP_MAX_OBJ_ID_LEN         32
+
+/* Max size + VLAN tag + extra PNET buffer */
+#define PROFINET_INFO_SIZE          0
+#define PBUF_FCS_SIZE               4
+#define PBUF_POOL_BUFSIZE           LWIP_MEM_ALIGN_SIZE(1518 + PBUF_FCS_SIZE  + PROFINET_INFO_SIZE)
 
 /* Stack sizes */
 #define DEFAULT_THREAD_STACKSIZE    1024
 #define TCPIP_THREAD_STACKSIZE      1768
+#define SNMP_STACK_SIZE             2300
 
 /* TCPIP thread priority */
 #define TCPIP_THREAD_PRIO           5
 
+/* SNMP server thread priority */
+#define SNMP_THREAD_PRIO            3
+
 /* Mailbox sizes */
 #define TCPIP_MBOX_SIZE             128
 #define DEFAULT_RAW_RECVMBOX_SIZE   5
@@ -106,6 +124,7 @@
  *    LWIP_DBG_OFF
  *    LWIP_DBG_ON
  */
+#define PBUF_DEBUG                  LWIP_DBG_OFF
 #define IP_DEBUG                    LWIP_DBG_OFF
 #define IGMP_DEBUG                  LWIP_DBG_OFF
 #define TCPIP_DEBUG                 LWIP_DBG_OFF
@@ -120,5 +139,7 @@
 #define TCP_FR_DEBUG                LWIP_DBG_OFF
 #define TCP_QLEN_DEBUG              LWIP_DBG_OFF
 #define TCP_RST_DEBUG               LWIP_DBG_OFF
+#define SNMP_DEBUG                  LWIP_DBG_OFF
+#define SNMP_MIB_DEBUG              LWIP_DBG_OFF
 
 #endif  /* __LWIPOPTS_H__ */
-- 
2.25.1

